src/app/clusters/closure-dimension-server/closure-dimension-cluster-logic.cpp

```diff
@@ -19,95 +19,162 @@
 
 } // namespace
 
-CHIP_ERROR ClusterLogic::Init(const ClusterConformance & conformance)
+CHIP_ERROR ClusterLogic::Init(const ClusterConformance & conformance, const ClusterInitParameters & clusterInitParameters)
 {
-    // TODO: Add logic to inital State form device or set it to fallback values
     VerifyOrReturnError(!mInitialized, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnError(conformance.Valid(), CHIP_ERROR_INVALID_DEVICE_DESCRIPTOR);
-    // TODO: Add logic to set inital mState to fallback values.
     mConformance = conformance;
+
+    // Need to set TranslationDirection, RotationAxis, ModulationType before Initilization of closure, as they should not be changed
+    // after Initalization.
+    if (conformance.HasFeature(Feature::kTranslation))
+    {
+        ReturnErrorOnFailure(SetTranslationDirection(clusterInitParameters.translationDirection));
+    }
+
+    if (conformance.HasFeature(Feature::kRotation))
+    {
+        ReturnErrorOnFailure(SetRotationAxis(clusterInitParameters.rotationAxis));
+    }
+
+    if (conformance.HasFeature(Feature::kModulation))
+    {
+        ReturnErrorOnFailure(SetModulationType(clusterInitParameters.modulationType));
+    }
+
     mInitialized = true;
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR ClusterLogic::SetCurrentState(const GenericCurrentStateStruct & currentState)
+// TODO: CurrentState should be QuietReporting.
+CHIP_ERROR ClusterLogic::SetCurrentState(const DataModel::Nullable<GenericCurrentStateStruct> & incomingCurrentState)
 {
-    // TODO : Q reporting for this attribute
-    // TODO: To implement the impact of the MoveTo command from the Closure Control cluster
+    assertChipStackLockedByCurrentThread();
 
     VerifyOrReturnError(mInitialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mState.currentState != incomingCurrentState, CHIP_NO_ERROR);
+
+    if (incomingCurrentState.IsNull())
+    {
+        // Mark CurrentState attribute as dirty only if value changes.
+        if (!mState.currentState.IsNull())
+        {
+            mState.currentState.SetNull();
+            mMatterContext.MarkDirty(Attributes::CurrentState::Id);
+        }
+        return CHIP_NO_ERROR;
+    }
 
-    if (currentState.position.HasValue())
+    // Validate the incoming Position value has valid input parameters and FeatureMap conformance.
+    if (incomingCurrentState.Value().position.HasValue())
     {
+        //  If the position member is present in the incoming CurrentState, we need to check if the Positioning
+        //  feature is supported by the closure. If the Positioning feature is not supported, return an error.
         VerifyOrReturnError(mConformance.HasFeature(Feature::kPositioning), CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);
-        VerifyOrReturnError(currentState.position.Value() <= PERCENT100THS_MAX_VALUE, CHIP_ERROR_INVALID_ARGUMENT);
+
+        // We don't need to check if values are present since the check was done above.
```

Comment:
With the updated structure, comment doesn't have the same value since the `HasValue`
```suggestion
```